### PR TITLE
feat(ext/net): Add multicasting APIs to DatagramConn (#10706)

### LINKED ISSUE #10706: Please expose multicast for UDP Datagram
Please expose multicast functions for Datagram connections. The functions can be found in std::net::UdpSocket. It will mostly likely require a new function to check if an address is either IPv4 or IPv6.

### PR DESCRIPTION
**Pull request by sgwilym**
Fixes #10706 and adds support for multicasting by adding `joinMulticastV4` and `joinMulticastV6` methods to `DatagramConn`:

```ts
const listener = Deno.listenDatagram({
  hostname: "224.0.0.251",
  port: 5353,
  transport: "udp",
  reuseAddress: true,
});

const mdnsMembership = await listener.joinMulticastV4("224.0.0.251", "127.0.0.1");

mdnsMembership.setLoopback(true); 
mdnsMembership.setTTL(50); // Only available on IPv4 multicast memberships.

console.log("Listening for a multicast message...");

const multicastMsg = await listener.receive();

console.log(multicastMsg);

mdnsMembership.leave(); // Leaves the multicast membership.

listener.close();
```

The API here is a middle-ground between Rust's exacting but very unforgiving APIs, and Node `dgram`'s very forgiving but quite vague APIs that doesn't really discern between IPv4 and IPv6 memberships.

In the proposed API, when you join a multicast group an object is returned with methods appropriate for the connection type and specific membership. For instance, a v4 membership will have `setTTL` method, whereas a v6 membership will not (because it can't). Another example is a `leave` method with the address and interface used to join the multicast group baked in.

If this API is too opinionated, I would probably go for an explicit API like Rust's which lets you shoot yourself in the foot.

I don't really know how tests for these methods could be implemented. Hypothetically it's just glue code, so... 🤷

### CODE REVIEW

**Comment by RangerMauve** [comment]
This would be very useful to have to enable local network service discovery use cases!

File: cli/tsc/dts/lib.deno.unstable.d.ts
Comment (aapoalas) [inline cli/tsc/dts/lib.deno.unstable.d.ts:814]: Not sure if it would be useful or just a hindrance, but these types could use string template literal types to define acceptable addresses.

File: ext/net/ops.rs
Comment (aapoalas) [inline ext/net/ops.rs:213]: Would probably be a bit more performant to use individual parameters for the address and interface: Iterating through objects is a bit of a performance pain.

**Review by aapoalas — CHANGES_REQUESTED** [review]
Hey, and thank you for the PR! Apologies that it's taken me so long to get around to reviewing it.

Code wise this looks good to me, and the API can be iterated upon (eg. would it after all be better to have a single API with overloaded address and network interface parameters). However, one thing that must be handled before merging is some basic sort of testing. At the very least the APIs themselves should be exercised, even if the sending and receiving of multicast messages isn't explicitly verified to work (through eg. another active listener).

File: cli/tsc/dts/lib.deno.unstable.d.ts
Comment (aapoalas) [inline cli/tsc/dts/lib.deno.unstable.d.ts:820]: Is the loopback status of a socket commonly changed while it's in use? Would it make sense to actually set this up as a parameter?

File: ext/net/ops.rs
Comment (aapoalas) [inline ext/net/ops.rs:320]: It seems like the `group_address` itself isn't actually needed, and instead the V4/V6 choice could just be a boolean parameter.

**Comment by x1127-dev** [comment]
Don't forget about `broadcast` please! I would be very grateful if we could have some sort of method [`setBroadcast`](https://nodejs.org/api/dgram.html#socketsetbroadcastflag) from `node:dgram`.

File: cli/tsc/dts/lib.deno.unstable.d.ts
Comment (sgwilym) [reply cli/tsc/dts/lib.deno.unstable.d.ts:820]: I wasn't sure about this either, but as nearly every API for multicasting I've seen has this as a method call, I felt it was the safest option. 

When I read about multicast loopback, it does seem hard to imagine that you'd be flicking it on and off all the time... but I'm hesitant to make it impossible just because of my limited imagination.

**Comment by sgwilym** [comment]
@x1127-dev I think that falls outside of this PR, but I do know that broadcasting is enabled by default when `listenDatagram` is called. You may find https://github.com/denoland/deno/pull/12817 interesting, as well as the PR which was merged instead of it (https://github.com/denoland/deno/pull/12860)

**Comment by sgwilym** [comment]
@aapoalas Also added basic API workout tests, and a test which sends and receives messages via two listeners (which I could only get a IPv4 version of working).

**Comment by x1127-dev** [comment]
> @x1127-dev I think that falls outside of this PR, but I do know that broadcasting is enabled by default when `listenDatagram` is called. You may find #12817 interesting, as well as the PR which was merged instead of it (#12860)

Thank you, it really works! It wasn't obvious to me that the broadcast flag is enabled by default after the Node

**Comment by aapoalas** [comment]
@sgwilym Hey, I'll try and get around to reviewing this ASAP but it will take me some time as I'm not available this weekend, weekdays are always busy, and next weekend I'm gone as well.

File: cli/tsc/dts/lib.deno.unstable.d.ts
Comment (aapoalas) [reply cli/tsc/dts/lib.deno.unstable.d.ts:814]: Yeah. IPv4 address could be typed (in a rough way) like this:
```ts
type IPv4Address = `${number}.${number}.${number}.${number}`;
```
This doesn't check that the numbers are between 0 and 255 but it's better than nothing.

That being said, this may just be a nuisance in some cases, like f.ex. reading an address from a JSON configuration file, as TS will just type the address as `string`.

File: cli/tsc/dts/lib.deno.unstable.d.ts
Comment (aapoalas) [reply cli/tsc/dts/lib.deno.unstable.d.ts:820]: Yeah, turning this into a parameter would make f.ex. Node compat probably impossible to achieve.

That being said... Maybe the loopback option could still be placed in some options object here or there. The most relevant place might be the Dgram creation itself, since this was for the entire socket right, and not just some membership?

We can then just leave this method here as an extra API that isn't really recommended in the general case.

File: cli/tsc/dts/lib.deno.unstable.d.ts
Comment (crowlKats) [reply cli/tsc/dts/lib.deno.unstable.d.ts:814]: we should just leave it as string

**Review by crowlKats — DISMISSED** [review]
Please move the complex return types to be interface declarations

File: cli/tsc/dts/lib.deno.unstable.d.ts
Comment (sgwilym) [reply cli/tsc/dts/lib.deno.unstable.d.ts]: I've added an extra optional loopback parameter to `Deno.listenDatagram`, and left the methods on the membership objects.

File: ext/net/01_net.js
Comment (aapoalas) [inline ext/net/01_net.js:265]: **Nitpick:** You could save a few bytes by doing `() => core.opAsync(...)` :)

File: ext/net/ops.rs
Comment (aapoalas) [inline ext/net/ops.rs:493]: Remove debug prints.

**Review by aapoalas — APPROVED** [review]
Minor nitpicks but looks good to me! I'd like someone from the core team to provide some feedback on if `joinMulticastV4` and `joinMulticastV6` are good or if we should unify to a `joinMulticast` that would internally slipt paths based on the address format.

Aside from those two minor points, I think this is ready to merge.

**Comment by sgwilym** [comment]
The main argument for separate v4/v6 methods is that TS then knows whether you got back a `MulticastV4Membership` or `MulticastV6Membership`, the latter having no `setTTL` function.

This could also be done with overloads, but this feels less ergonomic to me as it's not hugely obvious that joining an IPv6 multicast group uses a scope ID rather than an IP address, and it would be really easy to run `listener.joinMulticast("ff02::fb", "::1")` (where the two string params would indicate the IPv4 variant) and be confused why it's not working.

**Comment by aapoalas** [comment]
> The main argument for separate v4/v6 methods is that TS then knows whether you got back a `MulticastV4Membership` or `MulticastV6Membership`, the latter having no `setTTL` function.
> 
> This could also be done with overloads, but this feels less ergonomic to me as it's not hugely obvious that joining an IPv6 multicast group uses a scope ID rather than an IP address, and it would be really easy to run `listener.joinMulticast("ff02::fb", "::1")` (where the two string params would indicate the IPv4 variant) and be confused why it's not working.

Yeah, I agree. I'd just like a word from core as well :)

**Comment by aapoalas** [comment]
@sgwilym Needs a rebase again. Unfortunately we haven't gotten any feedback from the core team but I think that just means we're good enough to merge as an unstable API, so we can make changes in a followup if we find we need to.

**Pull request merged**
